processor/stream/stream_processor.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/elastic/apm-server/model/metric"
+	"github.com/elastic/apm-server/model/span"
+	"github.com/elastic/apm-server/model/transaction"
+	"github.com/elastic/apm-server/publish"
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/utility"
+	"github.com/elastic/apm-server/validation"
+)
+
+var (
+	ErrUnrecognizedObject = errors.New("did not recognize object type")
+)
+
+type StreamReader interface {
+	Read() (map[string]interface{}, error)
+	IsEOF() bool
+	LastLine() []byte
+}
+
+type StreamProcessor struct {
+	Tconfig transform.Config
+}
+
+const batchSize = 10
+
+var models = []struct {
+	key          string
+	schema       *jsonschema.Schema
+	modelDecoder func(interface{}, error) (transform.Transformable, error)
+}{
+	{
+		"transaction",
+		transaction.ModelSchema(),
+		transaction.V2DecodeEvent,
+	},
+	{
+		"span",
+		span.ModelSchema(),
+		span.V2DecodeEvent,
+	},
+	{
+		"metric",
+		metric.ModelSchema(),
+		metric.DecodeMetric,
+	},
+	{
+		"error",
+		er.ModelSchema(),
+		er.V2DecodeEvent,
+	},
+}
+
+func (v *StreamProcessor) readMetadata(reqMeta map[string]interface{}, reader StreamReader) (*metadata.Metadata, error) {
+	// first item is the metadata object
+	rawModel, err := reader.Read()
+	if err != nil {
+		if e, ok := err.(decoder.JSONDecodeError); ok || err == io.EOF {
+			return nil, &Error{
+				Type:     InvalidInputErrType,
+				Message:  e.Error(),
+				Document: string(reader.LastLine()),
+			}
+		}
+		return nil, err
+	}
+
+	rawMetadata, ok := rawModel["metadata"].(map[string]interface{})
+	if !ok {
+		return nil, &Error{
+			Type:     InvalidInputErrType,
+			Message:  ErrUnrecognizedObject.Error(),
+			Document: string(reader.LastLine()),
+		}
+	}
+
+	for k, v := range reqMeta {
+		utility.InsertInMap(rawMetadata, k, v.(map[string]interface{}))
+	}
+
+	// validate the metadata object against our jsonschema
+	err = validation.Validate(rawMetadata, metadata.ModelSchema())
+	if err != nil {
+		return nil, &Error{
+			Type:     InvalidInputErrType,
+			Message:  err.Error(),
+			Document: string(reader.LastLine()),
+		}
+	}
+
+	// create a metadata struct
+	metadata, err := metadata.DecodeMetadata(rawMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return metadata, nil
+}
+
+// handleRawModel validates and decodes a single json object into its struct form
+func (v *StreamProcessor) handleRawModel(rawModel map[string]interface{}) (transform.Transformable, error) {
+	for _, model := range models {
+		if entry, ok := rawModel[model.key]; ok {
+			err := validation.Validate(entry, model.schema)
+			if err != nil {
+				return nil, err
+			}
+
+			tr, err := model.modelDecoder(entry, err)
+			if err != nil {
+				return nil, err
+			}
+			return tr, nil
+		}
+	}
+	return nil, ErrUnrecognizedObject
+}
+
+// readBatch will read up to `batchSize` objects from the ndjson stream
+// it returns a slice of eventables, a serverResponse and a bool that indicates if we're at EOF.
```

Comment:
ah yes, that comment didn't get updated :) 

Returning the response instead of taking a reference would requires us to merge the two responses every time readBatch returns. There's no merging code atm. Do you think it's worth it?

Reply:
Errors should be edge cases. Thus, I don't expect a big performance impact one way or the other. I personally find it cleaner, but no strong opinion.

Reply:
I've updated the comment now. I prefer to leave the arguments for now, but happy to discuss again at a later point.